quinn-proto/src/connection/send_buffer.rs

```diff
@@ -31,18 +33,46 @@ impl SendBuffer {
 
     /// Append application data to the end of the stream
     pub fn write(&mut self, data: &[u8]) {
-        self.unacked.extend_from_slice(data);
+        let buf: Bytes = Bytes::from(data.to_owned());
+        self.unacked_segments.push_back(buf);
+        self.unacked_len += data.len();
```

Comment:
I'm concerned that this will drastically reduce performance for small writes. Could we allocate chunks with an explicit capacity (say, 4k) instead?

Reply:
I thought about it. Buy I feel like if someone wants to optimize small writes they might rather want to have a buffered writer as part of their application. Doing a "quinn syscall" will never be super efficient, due to Mutex boundaries crossed, waking up another task, etc.

Also this would then need to be done via `BytesMut` and splitting into already committed capacity and free one which might be used for another write. But `BytesMut` is currently broken, [and would in that case trigger an additional allocation](https://github.com/tokio-rs/bytes/issues/401).

Reply:
Writes are async wrt. task wakeups, so it's not clear to me that the overhead is *that* high.

> BytesMut is currently broken, and would in that case trigger an additional allocation.

That's disappointing, but I don't think it applies here, since we'd only be freezing after filling a chunk (presuming we get exactly the capacity we ask for, at least, which should be the case in practice).

Reply:
> Writes are async wrt. task wakeups, so it's not clear to me that the overhead is that high.

But it goes through the complete stack. Waking up connection (a different task), checking all states there, including manipulating timers, etc. I don't think it's that negligible - but it's currently also not that easy to get numbers.

> That's disappointing, but I don't think it applies here, since we'd only be freezing after filling a chunk (presuming we get exactly the capacity we ask for, at least, which should be the case in practice).

Well, you might need to have things frozen directly after the write if you want write queue is based on `Bytes` and things need to go out as soon as possible. Maybe an alternative is too keep the write queue based on `BytesMut`, but that doesn't play well into offering an interface to push `Bytes`.

Reply:
I did a small benchmark for the reuse approach with
```
pub fn write(&mut self, mut data: &[u8]) {
        let mut buf = if let Some(reserve) = self.reserve.take() {
            reserve
        } else {
            BytesMut::with_capacity(4*1024)
        };
        buf.extend_from_slice(data);

        let enq_buf = buf.split();
        if buf.capacity() > 10 {
            self.reserve = Some(buf);
        }

        assert_eq!(enq_buf.len(), data.len());
        self.unacked_segments.push_back(enq_buf);
        self.unacked_len += data.len();
        self.offset += data.len() as u64;
    }
```
    
which certainly isn't perfect but should od some reuse. It ended up being slower than the version without.
One thing which might come into play is the other weird part of `BytesMut` that the linked PR fixes, which is that it somehow rememebers about capacities from allocations which have long been split off, and then tries to allocate more than it should later on.

Btw: I think this affects large as well as small writes. Even if the write is initially large - at some point the flow control window will be filled. At that point the peer drains the stream, and from there on there will be continuously only small writes possible.

Anyway - I think this is better than the current approach where the full amount of queued bytes are shifted whenever some data is acked (a 1 MTU ack can cause a 16MB memcpy). Therefore I think merging it the way it is is and in the future looking for additional improvements is fine.

Reply:
Ah, good point re: ACKs. Fair enough!

Reply:
> Well, you might need to have things frozen directly after the write if you want write queue is based on Bytes and things need to go out as soon as possible.

It'd still give us an upper bound of a 4K (or whatever) realloc/copy vs. an entire flow control window.